### ISSUE TITLE
Accuracy

### ISSUE THREAD

**Issue by deepsea6034625** [mantasu/sunglasses-detector#13]
Hello. 
First of all, thanks for releasing this excellent project.

I tried to segment lenses from images.
But as you can see here, accuracy is not good.
Is there any way to improve the accuracy?

![BOSS_BOSS1044IT_807_00](https://github.com/mantasu/glasses-detector/assets/155721694/9e07fb72-ea12-4316-bf0e-0aa2a64dd340)
![BOSS_BOSS1150CS_4IN_03](https://github.com/mantasu/glasses-detector/assets/155721694/86a3883b-90cf-4654-a71c-1edaeee3f3eb)
![MMISSONI_MMI0153S_KB7_02](https://github.com/mantasu/glasses-detector/assets/155721694/0e2154ee-5149-4dde-a090-7a64b97c5418)
![BOSS_BOSS1150CS_4IN_07](https://github.com/mantasu/glasses-detector/assets/155721694/8e8613dd-ef90-4161-9d23-965316cc95aa)
![CARRERA_CARRERA3005S_NOA_03](https://github.com/mantasu/glasses-detector/assets/155721694/4b973d78-a0eb-4941-b552-c9dd35051409)
![MMISSONI_MMI0153S_807_02](https://github.com/mantasu/glasses-detector/assets/155721694/0b0d1bb0-f8a5-4259-89a1-7a1effae2325)
![CARRERA_CARRERA3005S_R1A_03](https://github.com/mantasu/glasses-detector/assets/155721694/f8810d4c-7f1c-418e-bcb6-80c6fa749561)
![BOSS_BOSS1150CS_807_00](https://github.com/mantasu/glasses-detector/assets/155721694/62fb1e40-e030-4961-aa99-9775bc4175e7)
![BOSS_BOSS0521S_003_01](https://github.com/mantasu/glasses-detector/assets/15572169
![BOSS_BOSS1150CS_FLL_01](https://github.com/mantasu/glasses-detector/assets/155721694/d6b48b76-5d23-41ce-b814-1db4441b89de)
![BOSS_BOSS1150CS_FLL_00](https://github.com/mantasu/glasses-detector/assets/155721694/3f77a110-4a65-49bf-8e0f-e5fd5f3b6910)
4/4b480698-bf43-4778-8398-d282be9bd8f4)
![BOSS_BOSS0521S_003_00](https://github.com/mantasu/glasses-detector/assets/155721694/ec5723e7-f6ad-4e2e-a801-cf3f25181ed1)

Hope to please help me.
Thanks again.

**Comment by deepsea6034625** [comment]
One more.

I used this code.

```
import numpy as np
from glasses_detector import GlassesSegmenter

# Initialize full segmentation model of size large
seg_full = GlassesSegmenter(kind='lenses', size='medium') # TODO: change medium to large

# Process the directory of images and save the results in various ways
seg_full.process_dir('images', 'results', format='img', batch_size=6, pbar=False)
```

**Comment by mantasu** [comment]
## Better Weights

Hi @deepsea920415, unfortunately, there's not much that can be done because the dataset is very small. Based on the data (standalone glasses) you are showing, I tried training a new `large` model which seems to have _slightly_ improved accuracy.

```python
segmenter = GlassesSegmenter(kind="lenses", size="large", weights="path/to/weights.pth")
```

Here are two weight files:
* Same lenses dataset, larger model: [segmentation-lenses-large.pth](https://www.kaggle.com/models/mantasu/glasses-lens-segmenter/pyTorch/segmentation-lenses-large)
* Fine-tuned on data without people: [segmentation-lenses-large-standalone-glasses.pth](https://www.kaggle.com/models/mantasu/glasses-lens-segmenter/pyTorch/segmentation-lenses-large-standalone-glasses)

## Square Inputs

Also, you might want to pad your image from both sides to make sure it is a square:

<details><summary><b>OPTION 1: preprocess image files</b></summary>

```python
import os
from PIL import Image
from glasses_segmenter import GlassesSegmenter

# Both directories must already exist
INPUT_DIR = "path/to/non_square"
PREPROCESS_DIR = "path/to/square"

def make_image_square(image_path):
    img = Image.open(image_path)
    width, height = img.size
    target_size = max(width, height)
    new_img = Image.new(img.mode, (target_size, target_size), img.getpixel((0,0)))
    left_padding = (target_size - width) // 2
    top_padding = (target_size - height) // 2
    new_img.paste(img, (left_padding, top_padding))
    return new_img
    
for filename in os.listdir(INPUT_DIR):
    img = make_image_square(os.path.join(INPUT_DIR, filename))
    img.save(os.path.join(PREPROCESS_DIR, filename))
    
segmenter = GlassesSegmenter(kind="lenses", size="large", weights="path/to/weights.pth")
segmenter.process_dir(PREPROCESS_DIR)
```

</details>

<details><summary><b>OPTION 2: extend GlassesSegmenter</b></summary>

```python
from glasses_detector import GlassesSegmenter

class MyGlassesSegmenter(GlassesSegmenter):
    @staticmethod
    def make_image_square(image_path):
        img = Image.open(image_path)
        width, height = img.size
        target_size = max(width, height)
        new_img = Image.new(img.mode, (target_size, target_size), img.getpixel((0,0)))
        left_padding = (target_size - width) // 2
        top_padding = (target_size - height) // 2
        new_img.paste(img, (left_padding, top_padding))
        return new_img
    
    def predict(self, image_paths, *args, **kwargs):
        # WARNING: image_paths must be a list of paths!
        images = [self.make_image_square(path) for path in image_paths]
        return super().predict(images, *args, **kwargs)
        
segmenter = MyGlassesSegmenter(kind="lenses", size="large", weights="path/to/weights.pth")
segmenter.process_dir("path/to/non_square")
```

</details>

## Further Improvements

To further improve accuracy, here are some further suggestions (they are, however, beyond the scope of this package):
* Use more data, and experiment with various data augmentation techniques
* Use dilation and erosion to smoothen the masks, e.g., close segmentation holes
* Subtract frames segmenter mask from full segmenter mask to get a "helper" mask
* Tune for specific segmenter architecture and hyperparaeters (e.g., using optuna, ray, pytorch lightning packages)
* Use cross-validation because the dataset is very small
* Train an ensemble of models, possibly with varying architectures

**Comment by deepsea6034625** [comment]
Thank you.
I'll test with them.

**Comment by deepsea6034625** [comment]
The accuracy is improved a lot.
But still some bad samples.

![CARRERA_CARRERA8901_I46_03](https://github.com/mantasu/glasses-detector/assets/155721694/cd3303d1-4276-4979-b1d8-f139cd2f5c45)
![CARRERA_CARRERA226_KJ1_03](https://github.com/mantasu/glasses-detector/assets/155721694/5937373d-c460-4bc1-8228-bde01d207b68)
![CARRERA_CARRERA291_R80_03](https://github.com/mantasu/glasses-detector/assets/155721694/ddaa9b72-4e82-4edf-9e67-916611bb6e57)
![CARRERA_CARRERA313_086_03](https://github.com/mantasu/glasses-detector/assets/155721694/dfb42f03-30aa-4ed9-8f7f-49cd4a636929)

But totally, I think overall accuracy is good.
If we add some more dataset, the accuracy will be better.
Thanks again.

**Comment by mantasu** [comment]
Yup, training on more data is the best way to go!
